## Maximum of a quadratic form on the positive octant of the unit sphere (general theory)

Let $Q$ be a real symmetric $(n\times n)$-matrix of a quadratic form $q: \mathbb{R}^n \to \mathbb{R}$ .

A principal $(d\times d)$-submatrix of $Q$ is formed by selecting a subset of $d$ rows and the same subset of columns from $Q$. 

A linear subspace $V \subset \mathbb{R}^d$ intersects the positive octant if there is a vector $\mathbf{v} \in V$ such that $v_i >0$ for all $i$, where $v_i$ are the components of $\mathbf{v} = (v_1, \ldots, v_d)$.

We say that an eigenvalue of a principal $(d \times d)$-submatrix of $Q$ is *feasible* if its eigenspace intersects the positive octant.

**Theorem 1:** The maximum value of the quadratic form $q: \mathbb{R}^n \to \mathbb{R}$ on the closed positive octant of the unit sphere
$$\Delta:= \{\mathbf{x} \in
 \mathbb{R}^n \,|\, x_i \geq 0\} \cap S^{n-1}$$ 
 is equal to the maximum of all feasible eigenvalues of all principal $(d \times d)$-submatrices of $Q$ for $1 \leq d \leq n$.

*Proof:* Starting with $d=1$, the maximum value of all diagonal elements of $Q$ is equal to the maximum $m_0$ of the quadratic form $q$ at the vertices
$\mathbf{e}_1, \ldots, \mathbf{e}_n$ of $\Delta$. 

Let $\delta$ be an edge of $\Delta$ connecting the vertices $\mathbf{e}_i$ and $\mathbf{e}_j$.
If $q$ attains a value $m_1 > m_0$ on $\delta$ , then $m_1$ is a feasible eigenvalue of the principal $(2\times 2)$-submatrix of $Q$ defined by $\{i,j\} \subset \{1, \ldots, n\}$. 

Taking faces of $\Delta$ of dimensions $2, 3, \ldots, n$ and arguing as above, the statement follows. $\square$

## The Hirzebruch quadratic form of the Non-Pappus matroid

In [1]:
%display latex

The matrix $Q$ of the Hirzebruch quadratic form of the Non-Pappus arrangement is given by 

In [2]:
Q = matrix([
    [-5, 1, 1, -2, 1, 1, 1, 1, -2],
    [1, -5, 1, 1, -2, 1, 1, -2, 1],
    [1, 1, -5, 1, 1, -2, -2, 1, 1],
    [-2, 1, 1, -5, 1, 1, 1, 1, -2],
    [1, -2, 1, 1, -5, 1, 1, -2, 1],
    [1, 1, -2, 1, 1, -5, -2, 1, 1],
    [1, 1, -2, 1, 1, -2, -2, -2, -2],
    [1, -2, 1, 1, -2, 1, -2, -2, -2],
    [-2, 1, 1, -2, 1, 1, -2, -2, -2]
]); Q

[-5  1  1 -2  1  1  1  1 -2]
[ 1 -5  1  1 -2  1  1 -2  1]
[ 1  1 -5  1  1 -2 -2  1  1]
[-2  1  1 -5  1  1  1  1 -2]
[ 1 -2  1  1 -5  1  1 -2  1]
[ 1  1 -2  1  1 -5 -2  1  1]
[ 1  1 -2  1  1 -2 -2 -2 -2]
[ 1 -2  1  1 -2  1 -2 -2 -2]
[-2  1  1 -2  1  1 -2 -2 -2]

In [3]:
Q.dimensions()

(9, 9)

In [4]:
Q.is_symmetric()

True

The characteristic polynomial of $Q$ is:

In [5]:
charpolQ = Q.characteristic_polynomial().factor(); charpolQ

(x + 6) * (x + 3)^4 * (x^2 + 9*x - 18)^2

The eigenvalues, multiplicities, and eigenspaces of $Q$ in increasing order are given as follows

In [6]:
eigQ = Q.eigenvectors_left()
eigQ.sort()

In [7]:
for e in eigQ:
    print(f"Eigenvalue: {e[0]}")
    print(f"Multiplicity: {e[-1]}")
    print("Eigenvectors: ")
    display(e[1])
    print()

Eigenvalue: -10.68465843842650?
Multiplicity: 2
Eigenvectors: 


[(1, 0, -1, 1, 0, -1, -0.5615528128088302?, 0, 0.5615528128088302?),
 (0, 1, -1, 0, 1, -1, -0.5615528128088302?, 0.5615528128088302?, 0)]


Eigenvalue: -6
Multiplicity: 1
Eigenvectors: 


[
(0, 0, 0, 0, 0, 0, 1, 1, 1)
]


Eigenvalue: -3
Multiplicity: 4
Eigenvectors: 


[
(1, 0, 0, 0, 1, 1, 0, 0, 0),
(0, 1, 0, 0, -1, 0, 0, 0, 0),
(0, 0, 1, 0, 0, -1, 0, 0, 0),
(0, 0, 0, 1, 1, 1, 0, 0, 0)
]


Eigenvalue: 1.684658438426491?
Multiplicity: 2
Eigenvectors: 


[(1, 0, -1, 1, 0, -1, 3.561552812808830?, 0, -3.561552812808830?),
 (0, 1, -1, 0, 1, -1, 3.561552812808830?, -3.561552812808830?, 0)]

In particular, the quadratic form $q: \mathbb{R}^9 \to \mathbb{R}$ is non-degenerate and has signature $(2,7)$.

#### The eigenvalues and eigenvector components are algebraic numbers

The lowest eigenvalue is $\approx -10.68$

In [8]:
lowest_eigenval_Q = eigQ[0][0]; lowest_eigenval_Q

-10.68465843842650?

Numbers ending in question marks are algebraic

In [9]:
type(lowest_eigenval_Q)

<class 'sage.rings.qqbar.AlgebraicNumber'>

The radical expression of `lowest_eigenval_Q` is

In [10]:
lowest_eigenval_Q.radical_expression()

-3/2*sqrt(17) - 9/2

Similarly, the highest eigenvalue is $\approx 1.68$

In [11]:
highest_eigenval_Q = eigQ[-1][0] 
display(highest_eigenval_Q.radical_expression())
print(f"= {highest_eigenval_Q}")

3/2*sqrt(17) - 9/2

= 1.684658438426491?


The lowest and highest eigenvalues of $Q$ are the roots of the polynomial $x^2+9x-18$ , the only factor of the characteristic polynomial that does not split over $\mathbb{Q}$

## Intersection of a subspace with the positive octant

Let $\mathbf{v}_1, \ldots, \mathbf{v}_k$ be vectors in $\mathbb{R}^n$.

We say that a linear combination $\mathbf{v} = \sum_{i=1}^k c_i \mathbf{v}$ is positive
if $\mathbf{v} \in \mathbb{R}^n_{>0}$ .

The next function returns the coefficients $c_1, \ldots, c_k$ of a positive linear combination.

It returns the empty list if no such combination exists.

In [12]:
def positive_linear_combination(vectors):
    '''
    Input: list of vectors
    Return: list of coefficients of a positive linear combination, 
    if there is no positive combination then it returns an empty list
    '''
    
    # Form a matrix with the input vectors as columns
    A = matrix(vectors).transpose()

    # n is the dimension of the vector space and k is the number of vectors
    n, k = A.dimensions()
    
    # Initialize the linear program
    p = MixedIntegerLinearProgram()
    c = p.new_variable(real=True)  # Coefficients can be arbitrary

    # Set a dummy objective function (optimize for 0)
    p.set_objective(0)

    # Add constraints to ensure that each component of the result is positive
    for i in range(n):
        p.add_constraint(sum(A[i, j] * c[j] for j in range(k)) >= 1)

    try:
        p.solve()
        d = p.get_values(c)
        return list(d.values())
    except:
        return []

Example:

In [13]:
v1 = vector([-4])

# List of vectors spanning the subspace
vectors = [v1]

positive_linear_combination(vectors)

[-0.25]

Example:

In [14]:
v1 = vector([1, -1, 0])
v2 = vector([0, -1, -1])

# List of vectors spanning the subspace
vectors = [v1, v2]

positive_linear_combination(vectors)

[1.0, -2.0]

Example:

In [15]:
v1 = vector([1, -1, 0])
v2 = vector([0, 2, 0])

# List of vectors spanning the subspace
vectors = [v1, v2]

positive_linear_combination(vectors)

[]

The next function checks if the span of a list of vectors intersects the positive octant

In [16]:
def intersects_positive_octant(vectors):
   
    p = positive_linear_combination(vectors)
    
    return len(p) > 0

Example:

In [17]:
v1 = vector([1, -1, 0])

# List of vectors spanning the subspace
vectors = [v1]

intersects_positive_octant(vectors)

False

Example:

In [18]:
v1 = vector([1, -1, 0])
v2 = vector([0, -1, -1])

vectors = [v1,v2]

intersects_positive_octant(vectors)

True

## Principal submatrices

The next function returns a list of all non-empty subsets of $[0, 1, \ldots,n-1]$

In [19]:
from itertools import combinations

def powerset(n):
    
    s = [_ for _ in range(n)]
    subsets = []
    for r in range(1,n+1):
        subsets += [list(x) for x in combinations(s,r)]
    return subsets

Example:

In [20]:
powerset(3)

[[0], [1], [2], [0, 1], [0, 2], [1, 2], [0, 1, 2]]

The next function returns a list (with no repeated elements) of the principal $(d \times d)$-submatrices of a given
$(n \times n)$-matrix for all $1 \leq d \leq n$

In [21]:
def principal_submatrices(matrix):
    n = matrix.dimensions()[0]
    multiindeces = powerset(n)
    submatrices = []
    for I in multiindeces:
        subm = matrix[I,I]
        if subm not in submatrices:
            submatrices.append(subm)
    return submatrices

In our case of interest, the matrix $Q$ has $216$ different principal submatrices

In [22]:
prin_submat_Q = principal_submatrices(Q)
len(prin_submat_Q)

216

The first $4$ submatrices of the list are:

In [23]:
prin_submat_Q[:4]

[
            [-5  1]  [-5 -2]
[-5], [-2], [ 1 -5], [-2 -5]
]

## Maximum of a quadratic form on the positive octant of the unit sphere

We implement Theorem 1 in the following function:

In [24]:
def max_positive_sphere(matrix):
    maxvalue = -oo
    submatrices = principal_submatrices(matrix)
    for m in submatrices:
        eig_m = m.left_eigenvectors()
        for e in eig_m:
            if e[0] > maxvalue and intersects_positive_octant(e[1]):
                maxvalue = e[0]
    return maxvalue

Example:

In [25]:
A = matrix([
    [1,2],
    [2,1]
]); A

[1 2]
[2 1]

In [26]:
max_positive_sphere(A)

3

**MAIN APPLICATION**

The maximum of the Hirzebruch quadratic form of the Non-Pappus matroid on the closed positive octant of the unit sphere is $-1$

In [27]:
max_positive_sphere(Q)

-1

## The maximum points

Now that we know that the maximum is $-1$, we search for the principal submatrices for which $-1$ is a feasible eigenvalue.

We collect the indices that define these principal submatrices and the corresponding $-1$-eigenvectors.

In [28]:
max_eigenvectors = []
max_indices = []
for I in powerset(9):
    submatrix = Q[I,I]
    eig_submatrix = submatrix.eigenvectors_left()
    for e in eig_submatrix:
        if e[0] == -1 and intersects_positive_octant(e[1]):
            max_eigenvectors.append(e[1])
            max_indices.append(I)

The eigenvectors and indices are:

In [29]:
max_eigenvectors

[[
 (1, 1, 1, 1, 4)
 ],
 [
 (1, 1, 1, 1, 4)
 ],
 [
 (1, 1, 1, 1, 4)
 ]]

In [30]:
max_indices

[[0, 1, 3, 4, 6], [0, 2, 3, 5, 7], [1, 2, 4, 5, 8]]

Every component of `max_eigenvectors` is a list consisting of a single vector, i.e., the $-1$-eigenspaces of the corrsponding principal submatrices are $1$ dimensional. This implies that the maximum points of the quadratic form on $\Delta = \{\mathbf{x} \in
 \mathbb{R}^9 \,|\, x_i \geq 0\} \cap S^{8}$ are isolated. In our case, there are $3$ maximum points. 

The correponding points in $\mathbb{R}^9$ are

In [31]:
points_R9 = []
for i in range(3):
    v = [0]*9
    positions = max_indices[i]
    values = [1,1,1,1,4]
    for j in range(5):
        k = positions[j]
        v[k] = values[j]
    v = vector(v)
    points_R9.append(v)

In [32]:
points_R9

[(1, 1, 0, 1, 1, 0, 4, 0, 0),
 (1, 0, 1, 1, 0, 1, 0, 4, 0),
 (0, 1, 1, 0, 1, 1, 0, 0, 4)]

These $3$ vectors have the same norm:

In [33]:
a = points_R9[0].norm(); a

2*sqrt(5)

The maximum points of the quadratic form $q$ on $\Delta$ are the corresponding unit vectors:
$$ \mathbf{x}_1 = \frac{1}{2\sqrt{5}} \cdot \left(1,\,1,\,0,\,1,\,1,\,0,\,4,\,0,\,0\right) $$
$$ \mathbf{x}_2 = \frac{1}{2\sqrt{5}} \cdot \left(1,\,0,\,1,\,1,\,0,\,1,\,0,\,4,\,0\right) $$ 
$$ \mathbf{x}_3 =  \frac{1}{2\sqrt{5}} \cdot \left(0,\,1,\,1,\,0,\,1,\,1,\,0,\,0,\,4\right) $$

We check that $q(\mathbf{x}_1) = q(\mathbf{x}_2) = q(\mathbf{x}_3) = -1$

In [34]:
for v in points_R9:
    unit_v = v / v.norm()
    q_of_unit_v = unit_v * Q * unit_v
    print(f"Value of quadratic form: {q_of_unit_v}")

Value of quadratic form: -1
Value of quadratic form: -1
Value of quadratic form: -1


Altogether, we have proved the main result of this notebook:

**Theorem 2:** The maximum value of $q$ on $\Delta$ is $-1$. 
Moreover, the maximum is attained at the $3$ points $\mathbf{x}_1, \mathbf{x}_2, \mathbf{x}_3$.

## Symmetry

The automorphism group of the Non-Pappus matroid is of order $6$. It is the dihedral group $D_3$ (the symmetry group of an equilateral triangle). 

As a result, the quadratic form $q$ is invariant under the action on $\mathbb{R}^9$ generated by $f$ and $g$ with

$$ f \left(x_1,\, x_2,\, x_3,\, x_4,\, x_5,\, x_6,\, x_7,\, x_8,\, x_9 \right) = \left(x_4,\, x_5,\, x_6,\, x_1,\, x_2,\, x_3,\, x_7,\, x_8,\, x_9 \right)$$
$$ g \left(x_1,\, x_2,\, x_3,\, x_4,\, x_5,\, x_6,\, x_7,\, x_8,\, x_9 \right) = \left(x_2,\, x_3,\, x_1,\, x_5,\, x_6,\, x_4,\, x_9,\, x_7,\, x_8 \right)$$

The maximum points $\mathbf{x}_1, \mathbf{x}_2, \mathbf{x}_3$ are invariant by $f$ and cyclically permuted by $g$.